learning-from-errors/overlapping-instances/article.md

```diff
@@ -0,0 +1,626 @@
+  Nothing               -> printMe x
+```
+</details>
+
+Similarly, we could change/hack the general instance to use `Typeable` and implement
+specialized behavior for `Int` in the general instance itself.
+
+```hs
+instance Typeable a => Printable a where
+  printMe a = case cast a of
+    (Just x :: Maybe Int) -> putStrLn "general instance for int"
+    Nothing               -> putStrLn "general instance"
+```
+
+<details>
+  <summary>Full code</summary>
+
+```hs
+{-# LANGUAGE FlexibleInstances   #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Main (main) where
+
+import Data.Typeable
+
+class Printable a where
+  printMe :: a -> IO ()
+
+instance Typeable a => Printable a where
+  printMe a = case cast a of
+    (Just x :: Maybe Int) -> putStrLn "general instance for int"
+    Nothing               -> putStrLn "general instance"
+
+main :: IO ()
+main = fn (5 :: Int)
+
+fn :: forall a. Typeable a => a -> IO ()
+fn x = printMe x
+```
+</details>
+
+## Flip-Flopping overlapping instances
+
+Here we look at a variant of this error that seemingly flip-flops when
+attempts are made to fix it. Let us look at a sample.
+
+```hs
+{-# LANGUAGE FlexibleInstances #-}
+
+module Main (main) where
+
+import Data.Typeable
+
+class Printable a where
+  printMe :: a -> IO ()
+
+instance Printable a where
+  printMe _ = putStrLn "General instance"
+
+instance Functor f => Printable (f a) where
+  printMe _ = putStrLn "Instance for a Functor"
+
+newtype MyType a = MyType a
+
+main :: IO ()
+main = printMe (MyType 5)
+
+```
+
+As expected, we get the overlapping instances error.
+
+```
+• Overlapping instances for Printable (MyType Char)
+    arising from a use of ‘printMe’
+  Matching instances:
+    instance Printable a -- Defined at app/Main.hs:10:10
+    instance Functor f => Printable (f a)
+      -- Defined at app/Main.hs:13:10
+```
+
+Since we saw a similar error in the last example and fixed it by
+removing one of the instances, it might appear that same could work here as well.
+
+And it also kind of makes sense. Since GHC is confused by two eligible instances, in all probability, removing one of them should fix it, right?
+
+So we try by removing the instance for `Printable a`.
+
+<details>
+  <summary>Show changed code</summary>
+
+```hs
+{-# LANGUAGE FlexibleInstances #-}
+
+module Main (main) where
+
+import Data.Typeable
+
+class Printable a where
+  printMe :: a -> IO ()
+
+instance Functor f => Printable (f a) where
+  printMe _ = putStrLn "Instance for a Functor"
+
+newtype MyType a = MyType a
+
+main :: IO ()
+main = printMe (MyType 'c')
+```
+</details>
+
+And when we re-compile, we get...
+
+```
+• No instance for (Functor MyType) arising from a use of ‘printMe’
+• In the expression: printMe (MyType 'c')
+  In an equation for ‘main’: main = printMe (MyType 'c')
+```
+
+To our great surprise, we find that removing one instance from the two eligible
+instances made GHC reject the remaining instance as well!
```

Comment:
To clarify, I agree that the code is rejected, but I disagree that the instance is rejected.

Reply:
I am changing it as follows. I have left the "..reject the remaining instance as well!" to leave it a bit dramatic. Let me know if you think if it's better to remove it.


```
To our great surprise, we find that removing one instance from the two eligible
instances seem to have made GHC reject the remaining instance as well!

But actually, if you actually bothered to look at the error, removing one
instance made GHC proceed with the remaining instance, but only a bit further.
Also the error that is thrown is not an overlapping instance error.

So anyway, we end up in the situation where GHC flip-flops between these two errors.
```

Reply:
> I just told you a wrong thing 

Well, it is not a wrong thing. There are two steps in the process by which GHC selects an instance to use at a function call site. AFAIU, both steps should be successful or else the instance cannot be used, and in my description, I consider this to be "rejection" of the instance.

But the "overlapping instance" error that this article discusses, gets thrown only in the first step. So from that point of view, it can be argued that rejection in the second step should not count as "rejection" in this context. This seem to be what Vald is pointing out.

Which is why I think it is not completely wrong, and depends on how you define "rejection" of an instance.

Besides, if we said "removing one instance causes GHC to throw another error", it is actually no big deal (I mean, it happens all the time; you fix one another, GHC moves to another error), and is not at all surprising. It does not convey the impression when someone actually come across this behavior (saying from my own experience). Because both errors are "rejecting" the instance as unsutable to be used where it is required.

Also FWIW, a bit later, this is included in the same section. So we also explain the details further down..

> But the remaining instance `Functor f => Printable (f a)`
> failed to work, but this failure happens at a later phase: when constraints are matched and
> after GHC has picked an instance. This is why we get a `No instance for Functor MyType`
> error instead of an overlapping instance error.

Reply:
This "flip-flopping" along with both errors being in the same "class" of error where GHC cannot find an instance, is mostly how I see (or used to see) the behavior. 

If it seem that this point of view is not easily relatable, I think we should probably rename the section.